EIPS/eip-1962.md

```diff
@@ -0,0 +1,190 @@
+- Final exponentiation. This part does not depend from the number of pairs. Length of the loop for this operation is different for different families and can either be calculated in from other parameters or specified in the interface. In any case, gas costs can be estimated upfront.
+
+Below there will be proposed interface for BLS12 family of curves with other interfaces being built in a similar manner.
+
+#### BLS12 family pairing operation interface
+
+- 1 byte, length of the `x` parameter.
+- BE encoding of parameter `x` (unsigned).
+- 1 byte, sign of parameter `x`. `0x00` is `+`, `0x01` is `-`.
+- 1 byte, specifies type of the pairing. Either `0x01` or `0x02`. This type depends of the curve and does not affect the gas cost.
+- structure of the extension tower:
+    - non-residue for Fp2 extension in BE encoding. This is just a single element in the base field and must be less than modulus.
+    - non-residue for Fp6 extension (that is "3 over 2"). This is an element in Fp2 and encoded using the rules from the begining of this EIP.
+- Encoding of the "A" parameter in Fp2 for the curve over the extension (Fp2).
+- Encoding of the "B" parameter in Fp2 for the curve over the extension (Fp2).
+
+For BLS12 family `x` parameter is enough for gas costs estimation. It's Hamming weight and bit length determines complexity of the Miller loop. Final exponentiation depends of the base field size.
+
+Proposed gas cost formula: `C_0 * quadratic(fe_length) + N * C_1 * fe_length * Hamming_and_bit_length(x)`
+
+#### Pairing operation interface for other families
+
+Complexity for BN family is also determined by the single scalar parameter, so interface will be similar to BLS12. For MNT4/6 and Cocks-Pinch curves there would be required to specify number of steps in the Miller loop explicitly in the interface.
+
+### Possible simplifications
+
+Due to high complexity of the proposed operations in the aspects of implementation, debuging and evaluation of the factors for gas costs it may be appropriate to either limit the set of curves at the moment of acceptance to some list and then extend it. Another approach (if it's technically possible) would be to have the "whilelist" contract that can be updated without consensus changes (w/o fork).
+
+In the case of limited set of curve the following set is proposed as a minimal:
+- BN254 curve from the current version of Ethereum
+- BN curve from DIZK with 2^32 roots of unity
+- BLS12-381 
+- BLS12-377 from ZEXE with large number of roots of unity
+- MNT4/6 cycle from the original [paper](https://eprint.iacr.org/2014/595.pdf). It's not too secure, but may give some freedom for experiments.
```

Comment:
Are you aware of the 128 bit security MNT curve?

Reply:
The ones from Coda are ~128-bit, and are defined over field size of about ~750 bits

Reply:
Forgot about them :) Yes, those may be put in if performance-wise such large module and expensive pairing (large Miller loop and expensive final exponentiation) would fit in reasonable ("allowed") execution time.